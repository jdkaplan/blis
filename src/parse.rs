use std::path::Path;

pub type ParseResult<T> = Result<T, (Option<T>, ParseErrors)>;

pub type ParseErrors = Vec<ParseError>;

#[derive(thiserror::Error, Debug)]
#[error("parse error")]
pub struct ParseError;

pub struct File {
    pub name: String,
    pub decls: Vec<Decl>,
}

pub enum Decl {
    Stmt(Stmt),
}

pub enum Stmt {
    Expr(Expr),
}

pub enum Expr {
    Literal(Literal),
}

pub enum Literal {
    U64(u64),
}

pub struct Parser;

impl Parser {
    pub fn parse(path: &Path, source: &str) -> ParseResult<File> {
        let mut errors = Vec::new();
        let mut decls = Vec::new();

        if source.trim() == "1" {
            let decl = Decl::Stmt(Stmt::Expr(Expr::Literal(Literal::U64(1))));
            decls.push(decl);
        } else {
            errors.push(ParseError);
        }

        let file = File {
            name: path.to_string_lossy().into(),
            decls,
        };

        if errors.is_empty() {
            Ok(file)
        } else {
            Err((Some(file), errors))
        }
    }
}
