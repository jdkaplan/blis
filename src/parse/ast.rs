use num_bigint::BigInt;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Program {
    pub decls: Vec<Declaration>,
}

#[derive(Debug, Clone, Serialize)]
pub enum Declaration {
    Type(Type),
    Func(Func),
    Let(Let),

    Statement(Statement),
}

#[derive(Debug, Clone, Serialize)]
pub enum Statement {
    Break(Break),
    Continue(Continue),
    Loop(Loop),
    Return(Return),

    Assignment(Assignment),
    Expression(Expression),
}

#[derive(Debug, Clone, Serialize)]
pub enum Expression {
    LogicOr(LogicOr),
}

impl Expression {
    pub fn self_terminating(&self) -> bool {
        match self {
            Expression::LogicOr(or) => or.self_terminating(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct LogicOr {
    pub first: LogicAnd,
    pub rest: Vec<LogicAnd>,
}

impl LogicOr {
    pub fn self_terminating(&self) -> bool {
        self.rest.is_empty() && self.first.self_terminating()
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct LogicAnd {
    pub first: Equality,
    pub rest: Vec<Equality>,
}

impl LogicAnd {
    pub fn self_terminating(&self) -> bool {
        self.rest.is_empty() && self.first.self_terminating()
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Equality {
    Value(Comparison),

    Eq(Box<Equality>, Comparison),
    Ne(Box<Equality>, Comparison),
}

impl Equality {
    pub fn self_terminating(&self) -> bool {
        match self {
            Equality::Value(comp) => comp.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Comparison {
    Value(Term),

    Ge(Box<Comparison>, Term),
    Gt(Box<Comparison>, Term),
    Le(Box<Comparison>, Term),
    Lt(Box<Comparison>, Term),
}

impl Comparison {
    pub fn self_terminating(&self) -> bool {
        match self {
            Comparison::Value(term) => term.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Term {
    Value(Factor),

    Add(Box<Term>, Factor),
    Sub(Box<Term>, Factor),
}

impl Term {
    pub fn self_terminating(&self) -> bool {
        match self {
            Term::Value(factor) => factor.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Factor {
    Value(Unary),

    Mul(Box<Factor>, Unary),
    Div(Box<Factor>, Unary),
    Rem(Box<Factor>, Unary),
}

impl Factor {
    pub fn self_terminating(&self) -> bool {
        match self {
            Factor::Value(unary) => unary.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Unary {
    Value(Call),

    Neg(Box<Unary>),
    Not(Box<Unary>),
}

impl Unary {
    pub fn self_terminating(&self) -> bool {
        match self {
            Unary::Value(call) => call.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Call {
    Value(Primary),

    Call(Box<Call>, Vec<Expression>),
    Index(Box<Call>, Box<Expression>),
    Field(Box<Call>, Identifier),
}

impl Call {
    pub fn self_terminating(&self) -> bool {
        match self {
            Call::Value(call) => call.self_terminating(),
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Primary {
    Block(Block),
    If(If),
    Group(Box<Expression>),

    Atom(Atom),
}

#[derive(Debug, Clone, Serialize)]
pub struct Object {
    pub ty: Identifier,
    pub fields: Vec<(Identifier, Expression)>,
}

#[derive(Debug, Clone, Serialize)]
pub struct List {
    pub items: Vec<Expression>,
}

impl Primary {
    pub fn self_terminating(&self) -> bool {
        match self {
            Primary::Block(_) | Primary::If(_) => true,
            Primary::Atom(_) | Primary::Group(_) => false,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Atom {
    Identifier(Identifier),
    Literal(Literal),
    Object(Object),
    List(List),
}

#[derive(Debug, Clone, Serialize)]
pub struct Assignment {
    pub place: Place,
    pub expr: Expression,
}

#[derive(Debug, Clone, Serialize)]
pub enum Place {
    Field(Call, Identifier),
    Identifier(Identifier),
    Index(Call, Box<Expression>),
}

#[derive(thiserror::Error, Debug, Clone, Serialize)]
#[error("cannot assign to `{:?}`", target)]
pub struct PlaceError {
    pub target: Expression,
}

impl TryFrom<Expression> for Place {
    type Error = PlaceError;

    fn try_from(expr: Expression) -> Result<Self, Self::Error> {
        let target = expr.clone();

        let Expression::LogicOr(expr) = expr;

        let LogicOr { first: expr, rest } = expr;
        if !rest.is_empty() {
            return Err(PlaceError { target });
        }

        let LogicAnd { first: expr, rest } = &expr;
        if !rest.is_empty() {
            return Err(PlaceError { target });
        }

        let Equality::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Comparison::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Term::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Factor::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        let Unary::Value(expr) = expr else {
            return Err(PlaceError { target });
        };

        match expr {
            Call::Value(expr) => {
                let Primary::Atom(expr) = expr else {
                    return Err(PlaceError { target });
                };

                let Atom::Identifier(ident) = expr else {
                    return Err(PlaceError { target });
                };

                // Prevent assigning to `self` to avoid implying that the receiver can be changed
                // that way.
                if ident.name == "self" {
                    return Err(PlaceError { target });
                }

                Ok(Place::Identifier(ident.clone()))
            }
            Call::Field(obj, ident) => Ok(Place::Field(*obj.clone(), ident.clone())),
            Call::Index(obj, key) => Ok(Place::Index(*obj.clone(), key.clone())),
            Call::Call(_, _) => Err(PlaceError { target }),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct Type {
    pub ident: Identifier,
    pub methods: Vec<Method>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Method {
    pub self_: bool,
    pub ident: Identifier,
    pub params: Vec<Identifier>,
    pub body: Block,
}

#[derive(Debug, Clone, Serialize)]
pub struct Func {
    pub ident: Identifier,
    pub params: Vec<Identifier>,
    pub body: Block,
}

#[derive(Debug, Clone, Serialize)]
pub struct Let {
    pub ident: Identifier,
    pub expr: Expression,
}

#[derive(Debug, Clone, Serialize)]
pub struct Loop {
    pub label: Option<Identifier>,
    pub body: Block,
}

#[derive(Debug, Clone, Serialize)]
pub struct Return {
    pub expr: Expression,
}

#[derive(Debug, Clone, Serialize)]
pub struct Break {
    pub label: Option<Identifier>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Continue {
    pub label: Option<Identifier>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Block {
    pub decls: Vec<Declaration>,
    pub expr: Option<Box<Expression>>,
}

#[derive(Debug, Clone, Serialize)]
pub struct If {
    pub condition: Box<Expression>,
    pub consequent: Block,
    pub alternative: Option<Block>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Identifier {
    pub name: String,
}

impl Identifier {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }

    pub fn empty() -> Self {
        Self {
            name: String::from(""),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum Literal {
    Nil,
    Boolean(bool),
    Integer(BigInt),
    Float(f64),
    String(String),
}
