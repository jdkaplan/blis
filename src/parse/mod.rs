pub mod ast;
pub mod lexer;
pub mod parser;
pub mod token;

pub use lexer::Lexer;
pub use parser::{ParseError, ParseErrors, Parser};
pub use token::{Lexeme, LexemeOwned, Token};
