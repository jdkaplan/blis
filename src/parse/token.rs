#[derive(Debug, Copy, Clone, Default)]
pub struct Lexeme<'source> {
    pub token: Token,
    pub text: &'source str,
    pub line: usize,
}

impl Lexeme<'_> {
    pub fn to_owned(&self) -> LexemeOwned {
        LexemeOwned {
            token: self.token,
            text: String::from(self.text),
            line: self.line,
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct LexemeOwned {
    pub token: Token,
    pub text: String,
    pub line: usize,
}

#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, Hash)]
pub enum Token {
    // Single-character tokens
    LeftParen,
    RightParen,
    LeftBracket,
    RightBracket,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    Percent,

    // One- or two-character tokens
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals
    Identifier,
    String,
    Integer,
    Float,

    // Keywords
    And,
    Break,
    Continue,
    Else,
    False,
    Func,
    If,
    Let,
    Loop,
    Nil,
    Object,
    Or,
    Return,
    True,

    // Synthetic tokens
    #[default]
    Error,
    Eof,
}
