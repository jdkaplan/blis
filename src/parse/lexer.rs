use tracing::instrument;

use crate::parse::{Lexeme, Token};

pub struct Lexer<'source> {
    source: &'source str,
    start: usize,
    current: usize,
    line: usize,
}

impl std::fmt::Debug for Lexer<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Lexer")
            .field("source", &format!("({} bytes)", self.source.bytes().len()))
            .field("start", &self.start)
            .field("current", &self.current)
            .field("line", &self.line)
            .finish()
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = Lexeme<'source>;

    fn next(&mut self) -> Option<Self::Item> {
        Some(self.scan_next())
    }
}

impl<'source> Lexer<'source> {
    pub fn new(source: &'source str) -> Self {
        Self {
            source,
            start: 0,
            current: 0,
            line: 1,
        }
    }

    #[instrument(level = "trace", ret)]
    pub fn scan_next(&mut self) -> Lexeme<'source> {
        self.skip_ignored();

        self.start = self.current;

        if self.at_eof() {
            return self.make(Token::Eof);
        }

        match self.advance() {
            // One character, no lookahead
            '(' => self.make(Token::LeftParen),
            ')' => self.make(Token::RightParen),
            '[' => self.make(Token::LeftBracket),
            ']' => self.make(Token::RightBracket),
            '{' => self.make(Token::LeftBrace),
            '}' => self.make(Token::RightBrace),
            ';' => self.make(Token::Semicolon),
            ',' => self.make(Token::Comma),
            '.' => self.make(Token::Dot),
            '-' => self.make(Token::Minus),
            '+' => self.make(Token::Plus),
            '/' => self.make(Token::Slash),
            '*' => self.make(Token::Star),
            '%' => self.make(Token::Percent),

            // One or two characters, 1 lookahead
            '!' => {
                if self.take('=') {
                    self.make(Token::BangEqual)
                } else {
                    self.make(Token::Bang)
                }
            }
            '=' => {
                if self.take('=') {
                    self.make(Token::EqualEqual)
                } else {
                    self.make(Token::Equal)
                }
            }
            '<' => {
                if self.take('=') {
                    self.make(Token::LessEqual)
                } else {
                    self.make(Token::Less)
                }
            }
            '>' => {
                if self.take('=') {
                    self.make(Token::GreaterEqual)
                } else {
                    self.make(Token::Greater)
                }
            }

            '"' => self.scan_string(),

            // TODO: Allow non-ASCII identifiers
            c if is_alpha(c) => self.scan_identifier(),
            c if is_digit(c) => self.scan_number(),

            _ => self.error("unexpected character"),
        }
    }

    fn skip_ignored(&mut self) {
        loop {
            match self.peek() {
                // Whitespace
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }

                // Comments
                '/' => {
                    if self.peek_next() == '/' {
                        while self.peek() != '\n' && !self.at_eof() {
                            self.advance();
                        }
                    } else {
                        return;
                    }
                }

                _ => return,
            }
        }
    }

    fn scan_identifier(&mut self) -> Lexeme<'source> {
        while is_alpha(self.peek()) || is_digit(self.peek()) {
            self.advance();
        }

        let text = &self.source[self.start..self.current];

        let token = keyword_or_identifier(text);
        self.make(token)
    }

    fn scan_number(&mut self) -> Lexeme<'source> {
        while is_digit(self.peek()) {
            self.advance();
        }

        let mut has_fraction = false;

        // Only consume the dot if there are also more digits afterward. Otherwise, it indicates a
        // method call rather than a decimal number.
        if self.peek() == '.' && is_digit(self.peek_next()) {
            self.advance(); // `.`
            has_fraction = true;

            while is_digit(self.peek()) {
                self.advance();
            }
        }

        if has_fraction {
            self.make(Token::Float)
        } else {
            self.make(Token::Integer)
        }
    }

    fn scan_string(&mut self) -> Lexeme<'source> {
        while self.peek() != '"' && !self.at_eof() {
            if self.peek() == '\n' {
                self.line += 1;
            }

            // There are two escape sequences that have to be handled here in the lexer
            // because they would otherwise cause strings to terminate early:
            //
            //  1. `\"` escapes the double-quote, allowing the string to continue.
            //  2. `\\` escapes the backslash, preventing `\\"` from escaping the quote.
            //
            // All other escapes are interpreted during parsing.
            if self.peek() == '\\' && matches!(self.peek_next(), '"' | '\\') {
                self.advance(); // peek \
                self.advance(); // peek_next
            } else {
                self.advance(); // peek
            }
        }

        if self.at_eof() {
            return self.error("unterminated string");
        }

        self.advance(); // peek `"`
        self.make(Token::String)
    }

    fn at_eof(&self) -> bool {
        self.peek() == '\0'
    }

    fn peek(&self) -> char {
        self.source[self.current..]
            .chars()
            .next()
            .unwrap_or_default()
    }

    fn peek_next(&self) -> char {
        if self.at_eof() {
            return '\0';
        }

        let mut chars = self.source[self.current..].chars();
        chars.next().unwrap(); // peek 0
        chars.next().unwrap_or_default() // peek 1
    }

    fn advance(&mut self) -> char {
        let c = self.peek();
        self.current += c.len_utf8();
        c
    }

    fn take(&mut self, expected: char) -> bool {
        if self.at_eof() {
            return false;
        }

        if self.peek() != expected {
            return false;
        }

        self.advance();
        true
    }

    fn make(&self, token: Token) -> Lexeme<'source> {
        let text = &self.source[self.start..self.current];

        Lexeme {
            token,
            text,
            line: self.line,
        }
    }

    fn error<'msg>(&self, msg: &'msg str) -> Lexeme<'msg> {
        Lexeme {
            token: Token::Error,
            text: msg,
            line: self.line,
        }
    }
}

fn is_alpha(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_digit(c: char) -> bool {
    c.is_ascii_digit()
}

fn keyword_or_identifier(text: &str) -> Token {
    match text {
        "and" => Token::And,
        "break" => Token::Break,
        "continue" => Token::Continue,
        "else" => Token::Else,
        "false" => Token::False,
        "func" => Token::Func,
        "if" => Token::If,
        "let" => Token::Let,
        "loop" => Token::Loop,
        "nil" => Token::Nil,
        "object" => Token::Object,
        "or" => Token::Or,
        "return" => Token::Return,
        "true" => Token::True,
        _ => Token::Identifier,
    }
}
