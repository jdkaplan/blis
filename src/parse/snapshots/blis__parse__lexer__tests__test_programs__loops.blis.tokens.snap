---
source: src/parse/lexer.rs
description: "func main() {\n    let x = 0;\n    loop twice {\n        if x >= 2 {\n            break twice;\n        }\n\n        println(\"twice before\", x);\n        x = x + 1;\n        println(\"twice after \", x);\n    }\n    println(\"out of twice\", x);\n\n    loop outer {\n        if x < 0 {\n            println(\"break outer\", x);\n            break;\n        }\n\n        loop inner {\n            loop {\n                break;\n            }\n\n            if x > 2 {\n                println(\"break inner\", x);\n                break;\n            }\n\n            println(\"inner before\", x);\n            x = x + 1;\n            println(\"inner after \", x);\n        }\n        println(\"out of inner\", x);\n        // x = 3\n\n        if x > 2 {\n            x = -1;\n            continue;\n        }\n\n        println(\"never happens\");\n        x = x - 1;\n    }\n    println(\"out of outer\", x);\n}\n\nmain();\n"
input_file: test_programs/loops.blis
---
[
  Lexeme(
    token: Func,
    text: "func",
    line: 1,
  ),
  Lexeme(
    token: Identifier,
    text: "main",
    line: 1,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 1,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 1,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 1,
  ),
  Lexeme(
    token: Let,
    text: "let",
    line: 2,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 2,
  ),
  Lexeme(
    token: Equal,
    text: "=",
    line: 2,
  ),
  Lexeme(
    token: Integer,
    text: "0",
    line: 2,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 2,
  ),
  Lexeme(
    token: Loop,
    text: "loop",
    line: 3,
  ),
  Lexeme(
    token: Identifier,
    text: "twice",
    line: 3,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 3,
  ),
  Lexeme(
    token: If,
    text: "if",
    line: 4,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 4,
  ),
  Lexeme(
    token: GreaterEqual,
    text: ">=",
    line: 4,
  ),
  Lexeme(
    token: Integer,
    text: "2",
    line: 4,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 4,
  ),
  Lexeme(
    token: Break,
    text: "break",
    line: 5,
  ),
  Lexeme(
    token: Identifier,
    text: "twice",
    line: 5,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 5,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 6,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 8,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 8,
  ),
  Lexeme(
    token: String,
    text: "\"twice before\"",
    line: 8,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 8,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 8,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 8,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 8,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 9,
  ),
  Lexeme(
    token: Equal,
    text: "=",
    line: 9,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 9,
  ),
  Lexeme(
    token: Plus,
    text: "+",
    line: 9,
  ),
  Lexeme(
    token: Integer,
    text: "1",
    line: 9,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 9,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 10,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 10,
  ),
  Lexeme(
    token: String,
    text: "\"twice after \"",
    line: 10,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 10,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 10,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 10,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 10,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 11,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 12,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 12,
  ),
  Lexeme(
    token: String,
    text: "\"out of twice\"",
    line: 12,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 12,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 12,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 12,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 12,
  ),
  Lexeme(
    token: Loop,
    text: "loop",
    line: 14,
  ),
  Lexeme(
    token: Identifier,
    text: "outer",
    line: 14,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 14,
  ),
  Lexeme(
    token: If,
    text: "if",
    line: 15,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 15,
  ),
  Lexeme(
    token: Less,
    text: "<",
    line: 15,
  ),
  Lexeme(
    token: Integer,
    text: "0",
    line: 15,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 15,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 16,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 16,
  ),
  Lexeme(
    token: String,
    text: "\"break outer\"",
    line: 16,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 16,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 16,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 16,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 16,
  ),
  Lexeme(
    token: Break,
    text: "break",
    line: 17,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 17,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 18,
  ),
  Lexeme(
    token: Loop,
    text: "loop",
    line: 20,
  ),
  Lexeme(
    token: Identifier,
    text: "inner",
    line: 20,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 20,
  ),
  Lexeme(
    token: Loop,
    text: "loop",
    line: 21,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 21,
  ),
  Lexeme(
    token: Break,
    text: "break",
    line: 22,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 22,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 23,
  ),
  Lexeme(
    token: If,
    text: "if",
    line: 25,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 25,
  ),
  Lexeme(
    token: Greater,
    text: ">",
    line: 25,
  ),
  Lexeme(
    token: Integer,
    text: "2",
    line: 25,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 25,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 26,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 26,
  ),
  Lexeme(
    token: String,
    text: "\"break inner\"",
    line: 26,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 26,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 26,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 26,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 26,
  ),
  Lexeme(
    token: Break,
    text: "break",
    line: 27,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 27,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 28,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 30,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 30,
  ),
  Lexeme(
    token: String,
    text: "\"inner before\"",
    line: 30,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 30,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 30,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 30,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 30,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 31,
  ),
  Lexeme(
    token: Equal,
    text: "=",
    line: 31,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 31,
  ),
  Lexeme(
    token: Plus,
    text: "+",
    line: 31,
  ),
  Lexeme(
    token: Integer,
    text: "1",
    line: 31,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 31,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 32,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 32,
  ),
  Lexeme(
    token: String,
    text: "\"inner after \"",
    line: 32,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 32,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 32,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 32,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 32,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 33,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 34,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 34,
  ),
  Lexeme(
    token: String,
    text: "\"out of inner\"",
    line: 34,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 34,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 34,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 34,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 34,
  ),
  Lexeme(
    token: If,
    text: "if",
    line: 37,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 37,
  ),
  Lexeme(
    token: Greater,
    text: ">",
    line: 37,
  ),
  Lexeme(
    token: Integer,
    text: "2",
    line: 37,
  ),
  Lexeme(
    token: LeftBrace,
    text: "{",
    line: 37,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 38,
  ),
  Lexeme(
    token: Equal,
    text: "=",
    line: 38,
  ),
  Lexeme(
    token: Minus,
    text: "-",
    line: 38,
  ),
  Lexeme(
    token: Integer,
    text: "1",
    line: 38,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 38,
  ),
  Lexeme(
    token: Continue,
    text: "continue",
    line: 39,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 39,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 40,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 42,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 42,
  ),
  Lexeme(
    token: String,
    text: "\"never happens\"",
    line: 42,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 42,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 42,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 43,
  ),
  Lexeme(
    token: Equal,
    text: "=",
    line: 43,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 43,
  ),
  Lexeme(
    token: Minus,
    text: "-",
    line: 43,
  ),
  Lexeme(
    token: Integer,
    text: "1",
    line: 43,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 43,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 44,
  ),
  Lexeme(
    token: Identifier,
    text: "println",
    line: 45,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 45,
  ),
  Lexeme(
    token: String,
    text: "\"out of outer\"",
    line: 45,
  ),
  Lexeme(
    token: Comma,
    text: ",",
    line: 45,
  ),
  Lexeme(
    token: Identifier,
    text: "x",
    line: 45,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 45,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 45,
  ),
  Lexeme(
    token: RightBrace,
    text: "}",
    line: 46,
  ),
  Lexeme(
    token: Identifier,
    text: "main",
    line: 48,
  ),
  Lexeme(
    token: LeftParen,
    text: "(",
    line: 48,
  ),
  Lexeme(
    token: RightParen,
    text: ")",
    line: 48,
  ),
  Lexeme(
    token: Semicolon,
    text: ";",
    line: 48,
  ),
  Lexeme(
    token: Eof,
    text: "",
    line: 49,
  ),
]
